src/openms/include/OpenMS/ANALYSIS/ID/IDScoreSwitcherAlgorithm.h

```diff
@@ -0,0 +1,237 @@
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Julianus Pfeuffer $
+// $Authors: Julianus Pfeuffer $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/CONCEPT/Exception.h>
+#include <OpenMS/KERNEL/ConsensusMap.h>
+
+#include <vector>
+#include <set>
+
+using namespace std;
+
+namespace OpenMS
+{
+
+  class OPENMS_DLLAPI IDScoreSwitcherAlgorithm:
+    public DefaultParamHandler
+  {
+  public:
+    IDScoreSwitcherAlgorithm ();
+
+    enum class ScoreType
+    {
+      RAW,
+      RAW_EVAL,
+      PP,
+      PEP,
+      FDR,
+      QVAL,
+    };
+
+    template <typename IDType>
+    void switchScores(IDType& id, Size& counter)
+    {
+      for (typename std::vector<typename IDType::HitType>::iterator hit_it = id.getHits().begin();
+           hit_it != id.getHits().end(); ++hit_it, ++counter)
+      {
+        if (!hit_it->metaValueExists(new_score_))
+        {
+          String msg = "Meta value '" + new_score_ + "' not found for " +
+                       describeHit_(*hit_it);
+          throw Exception::MissingInformation(__FILE__, __LINE__,
+                                              OPENMS_PRETTY_FUNCTION, msg);
+        }
+
+        String old_score_meta = (old_score_.empty() ? id.getScoreType() :
+                                 old_score_);
+        DataValue dv = hit_it->getMetaValue(old_score_meta);
+        if (!dv.isEmpty()) // meta value for old score already exists
+        {
+          if (fabs((double(dv) - hit_it->getScore()) * 2.0 /
+                   (double(dv) + hit_it->getScore())) > tolerance_)
+          {
+            String msg = "Meta value '" + old_score_meta + "' already exists "
+                                                           "with a conflicting value for " + describeHit_(*hit_it);
+            throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION,
+                                          msg, dv.toString());
+          } // else: values match, nothing to do
+        }
+        else
+        {
+          hit_it->setMetaValue(old_score_meta, hit_it->getScore());
+        }
+        hit_it->setScore(hit_it->getMetaValue(new_score_));
+      }
+      id.setScoreType(new_score_type_);
+      id.setHigherScoreBetter(higher_better_);
+    }
+
+    /// Looks at the first Hit of the given @p id and according to the given @p type ,
+    /// deduces a fitting score and score direction to be switched to.
+    /// Then tries to switch all hits.
+    void switchToGeneralScoreType(vector<PeptideIdentification>& id, ScoreType type, Size& counter)
```

Comment:
this sounds a bit like what the MS obo models with parent CV nodes of certain score types. I guess MS obo is not complete here otherwise it could be a good way to document that

Reply:
Yes, I looked into it and it is far from complete. Also, it would be quite a large change and is more something for the new ID structure.